// @svrooij/sonos/lib/services/alarm-clock.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlarmClockServiceBase = void 0;
/**
 * Sonos AlarmClock service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Control the sonos alarms and times
 *
 * @export
 * @class AlarmClockServiceBase
 * @extends {BaseService}
 */
class AlarmClockServiceBase extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'AlarmClock';
        this.controlUrl = '/AlarmClock/Control';
        this.eventSubUrl = '/AlarmClock/Event';
        this.scpUrl = '/xml/AlarmClock1.xml';
        /**
         * Default errors and service specific errors
         *
         * @type {SonosUpnpError[]}
         * @remarks See https://svrooij.io/sonos-api-docs/#manual-documentation-file
         */
        this.errors = [
            ...sonos_upnp_errors_1.default.defaultErrors,
            { code: 801, description: 'Duplicate alarm time' },
        ];
    }
    // #region actions
    /**
     * Create a single alarm, all properties are required
     *
     * @param {string} input.StartLocalTime - The start time as `hh:mm:ss`
     * @param {string} input.Duration - The duration as `hh:mm:ss`
     * @param {string} input.Recurrence - Repeat this alarm on [ 'ONCE' / 'WEEKDAYS' / 'WEEKENDS' / 'DAILY' ]
     * @param {boolean} input.Enabled - Alarm enabled after creation
     * @param {string} input.RoomUUID - The UUID of the speaker you want this alarm for
     * @param {string} input.ProgramURI - The sound uri
     * @param {Track | string} input.ProgramMetaData - The sound metadata, can be empty string
     * @param {PlayMode} input.PlayMode - Alarm play mode [ 'NORMAL' / 'REPEAT_ALL' / 'SHUFFLE_NOREPEAT' / 'SHUFFLE' ]
     * @param {number} input.Volume - Volume between 0 and 100
     * @param {boolean} input.IncludeLinkedZones - Should grouped players also play the alarm?
     */
    async CreateAlarm(input) { return await this.SoapRequestWithBody('CreateAlarm', input); }
    /**
     * Delete an alarm
     *
     * @param {number} input.ID - The Alarm ID from ListAlarms
     */
    async DestroyAlarm(input) { return await this.SoapRequestWithBodyNoResponse('DestroyAlarm', input); }
    async GetDailyIndexRefreshTime() { return await this.SoapRequest('GetDailyIndexRefreshTime'); }
    async GetFormat() { return await this.SoapRequest('GetFormat'); }
    async GetHouseholdTimeAtStamp(input) { return await this.SoapRequestWithBody('GetHouseholdTimeAtStamp', input); }
    async GetTimeNow() { return await this.SoapRequest('GetTimeNow'); }
    async GetTimeServer() { return await this.SoapRequest('GetTimeServer'); }
    async GetTimeZone() { return await this.SoapRequest('GetTimeZone'); }
    async GetTimeZoneAndRule() { return await this.SoapRequest('GetTimeZoneAndRule'); }
    async GetTimeZoneRule(input) { return await this.SoapRequestWithBody('GetTimeZoneRule', input); }
    /**
     * Get the AlarmList as XML
     * @remarks Some libraries also provide a ListAndParseAlarms where the alarm list xml is parsed
     */
    async ListAlarms() { return await this.SoapRequest('ListAlarms'); }
    async SetDailyIndexRefreshTime(input) { return await this.SoapRequestWithBodyNoResponse('SetDailyIndexRefreshTime', input); }
    async SetFormat(input) { return await this.SoapRequestWithBodyNoResponse('SetFormat', input); }
    async SetTimeNow(input) { return await this.SoapRequestWithBodyNoResponse('SetTimeNow', input); }
    async SetTimeServer(input) { return await this.SoapRequestWithBodyNoResponse('SetTimeServer', input); }
    async SetTimeZone(input) { return await this.SoapRequestWithBodyNoResponse('SetTimeZone', input); }
    /**
     * Update an alarm, all parameters are required.
     *
     * @param {number} input.ID - The ID of the alarm see ListAlarms
     * @param {string} input.StartLocalTime - The start time as `hh:mm:ss`
     * @param {string} input.Duration - The duration as `hh:mm:ss`
     * @param {string} input.Recurrence - Repeat this alarm on [ 'ONCE' / 'WEEKDAYS' / 'WEEKENDS' / 'DAILY' ]
     * @param {boolean} input.Enabled - Alarm enabled after creation
     * @param {string} input.RoomUUID - The UUID of the speaker you want this alarm for
     * @param {string} input.ProgramURI - The sound uri
     * @param {Track | string} input.ProgramMetaData - The sound metadata, can be empty string
     * @param {PlayMode} input.PlayMode - Alarm play mode [ 'NORMAL' / 'REPEAT_ALL' / 'SHUFFLE_NOREPEAT' / 'SHUFFLE' ]
     * @param {number} input.Volume - Volume between 0 and 100
     * @param {boolean} input.IncludeLinkedZones - Should grouped players also play the alarm?
     * @remarks Some libraries support PatchAlarm where you can update a single parameter
     */
    async UpdateAlarm(input) { return await this.SoapRequestWithBodyNoResponse('UpdateAlarm', input); }
    // #endregion
    responseProperties() {
        return {
            AssignedID: 'number',
            CurrentDailyIndexRefreshTime: 'string',
            CurrentTimeFormat: 'string',
            CurrentDateFormat: 'string',
            HouseholdUTCTime: 'string',
            CurrentUTCTime: 'string',
            CurrentLocalTime: 'string',
            CurrentTimeZone: 'string',
            CurrentTimeGeneration: 'number',
            CurrentTimeServer: 'string',
            Index: 'number',
            AutoAdjustDst: 'boolean',
            TimeZone: 'string',
            CurrentAlarmList: 'string',
            CurrentAlarmListVersion: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            AlarmListVersion: 'string',
            DailyIndexRefreshTime: 'string',
            DateFormat: 'string',
            TimeFormat: 'string',
            TimeGeneration: 'number',
            TimeServer: 'string',
            TimeZone: 'string',
        };
    }
}
exports.AlarmClockServiceBase = AlarmClockServiceBase;
