// @svrooij/sonos/lib/services/alarm-clock.service.extension.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlarmClockService = void 0;
const array_helper_1 = __importDefault(require("../helpers/array-helper"));
const metadata_helper_1 = __importDefault(require("../helpers/metadata-helper"));
const xml_helper_1 = __importDefault(require("../helpers/xml-helper"));
const alarm_clock_service_1 = require("./alarm-clock.service");
/**
 * Extended AlarmClockService
 *
 * @export
 * @class AlarmClockService
 * @extends {AlarmClockServiceBase}
 */
class AlarmClockService extends alarm_clock_service_1.AlarmClockServiceBase {
    /**
     * Get a parsed list of all alarms.
     *
     * @returns {Promise<Alarm[]>}
     * @memberof SonosDevice
     */
    async ListAndParseAlarms() {
        const alarmList = await super.ListAlarms();
        if (alarmList.CurrentAlarmList === undefined || alarmList.CurrentAlarmList === '') {
            return [];
        }
        const parsedList = xml_helper_1.default.DecodeAndParseXml(alarmList.CurrentAlarmList, '');
        const alarms = array_helper_1.default.ForceArray(parsedList.Alarms.Alarm);
        const results = [];
        alarms.forEach((alarm) => {
            if (alarm.ID !== undefined) {
                results.push({
                    Duration: alarm.Duration,
                    Enabled: alarm.Enabled === '1',
                    ID: parseInt(alarm.ID, 10),
                    IncludeLinkedZones: alarm.IncludeLinkedZones === '1',
                    PlayMode: alarm.PlayMode,
                    ProgramMetaData: metadata_helper_1.default.ParseDIDLTrack(xml_helper_1.default.DecodeAndParseXml(alarm.ProgramMetaData), this.host, this.port),
                    ProgramURI: xml_helper_1.default.DecodeTrackUri(alarm.ProgramURI),
                    Recurrence: alarm.Recurrence,
                    RoomUUID: alarm.RoomUUID,
                    StartLocalTime: alarm.StartTime,
                    Volume: parseInt(alarm.Volume, 10),
                });
            }
        });
        return results;
    }
    /**
     * Patch a single alarm. Only the ID and one property you want to change are required.
     *
     * @param {PatchAlarm} [options]
     * @param {number} options.ID The ID of the alarm to update
     * @param {string | undefined} options.StartLocalTime The time the alarm has to start 'hh:mm:ss'
     * @param {string | undefined} options.Duration The duration of the alarm 'hh:mm:ss'
     * @param {string | undefined} options.Recurrence What should the recurrence be ['DAILY','ONCE','WEEKDAYS']
     * @param {boolean | undefined} options.Enabled Should this alarm be enabled
     * @param {PlayMode | undefined} options.PlayMode What playmode should be used
     * @param {number | undefined} options.Volume The requested alarm volume
     * @returns {Promise<boolean>}
     * @memberof SonosDevice
     */
    async PatchAlarm(options) {
        this.debug('AlarmPatch(%o)', options);
        const alarms = await this.ListAndParseAlarms();
        const alarm = alarms.find((a) => a.ID === options.ID);
        if (alarm === undefined) {
            throw new Error(`Alarm with ID ${options.ID} not found`);
        }
        if (options.Duration !== undefined)
            alarm.Duration = options.Duration;
        if (options.Enabled !== undefined)
            alarm.Enabled = options.Enabled;
        if (options.PlayMode !== undefined)
            alarm.PlayMode = options.PlayMode;
        if (options.Recurrence !== undefined)
            alarm.Recurrence = options.Recurrence;
        if (options.StartLocalTime !== undefined)
            alarm.StartLocalTime = options.StartLocalTime;
        if (options.Volume !== undefined)
            alarm.Volume = options.Volume;
        return await this.UpdateAlarm(alarm);
    }
}
exports.AlarmClockService = AlarmClockService;
