// @svrooij/sonos/lib/services/audio-in.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioInService = void 0;
/**
 * Sonos AudioIn service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Control line in
 *
 * @export
 * @class AudioInService
 * @extends {BaseService}
 */
class AudioInService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'AudioIn';
        this.controlUrl = '/AudioIn/Control';
        this.eventSubUrl = '/AudioIn/Event';
        this.scpUrl = '/xml/AudioIn1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async GetAudioInputAttributes() { return await this.SoapRequest('GetAudioInputAttributes'); }
    async GetLineInLevel() { return await this.SoapRequest('GetLineInLevel'); }
    async SelectAudio(input) { return await this.SoapRequestWithBodyNoResponse('SelectAudio', input); }
    async SetAudioInputAttributes(input) { return await this.SoapRequestWithBodyNoResponse('SetAudioInputAttributes', input); }
    async SetLineInLevel(input) { return await this.SoapRequestWithBodyNoResponse('SetLineInLevel', input); }
    async StartTransmissionToGroup(input) { return await this.SoapRequestWithBody('StartTransmissionToGroup', input); }
    async StopTransmissionToGroup(input) { return await this.SoapRequestWithBodyNoResponse('StopTransmissionToGroup', input); }
    // #endregion
    responseProperties() {
        return {
            CurrentName: 'string',
            CurrentIcon: 'string',
            CurrentLeftLineInLevel: 'number',
            CurrentRightLineInLevel: 'number',
            CurrentTransportSettings: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            AudioInputName: 'string',
            Icon: 'string',
            LeftLineInLevel: 'number',
            LineInConnected: 'boolean',
            Playing: 'boolean',
            RightLineInLevel: 'number',
        };
    }
}
exports.AudioInService = AudioInService;
