// @svrooij/sonos/lib/services/connection-manager.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionManagerService = void 0;
/**
 * Sonos ConnectionManager service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Services related to connections and protocols
 *
 * @export
 * @class ConnectionManagerService
 * @extends {BaseService}
 */
class ConnectionManagerService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'ConnectionManager';
        this.controlUrl = '/MediaRenderer/ConnectionManager/Control';
        this.eventSubUrl = '/MediaRenderer/ConnectionManager/Event';
        this.scpUrl = '/xml/ConnectionManager1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async GetCurrentConnectionIDs() { return await this.SoapRequest('GetCurrentConnectionIDs'); }
    async GetCurrentConnectionInfo(input) { return await this.SoapRequestWithBody('GetCurrentConnectionInfo', input); }
    async GetProtocolInfo() { return await this.SoapRequest('GetProtocolInfo'); }
    // #endregion
    responseProperties() {
        return {
            ConnectionIDs: 'string',
            RcsID: 'number',
            AVTransportID: 'number',
            ProtocolInfo: 'string',
            PeerConnectionManager: 'string',
            PeerConnectionID: 'number',
            Direction: 'string',
            Status: 'string',
            Source: 'string',
            Sink: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            CurrentConnectionIDs: 'string',
            SinkProtocolInfo: 'string',
            SourceProtocolInfo: 'string',
        };
    }
}
exports.ConnectionManagerService = ConnectionManagerService;
