// @svrooij/sonos/lib/services/content-directory.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentDirectoryServiceBase = void 0;
/**
 * Sonos ContentDirectory service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Browse for local content
 *
 * @export
 * @class ContentDirectoryServiceBase
 * @extends {BaseService}
 */
class ContentDirectoryServiceBase extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'ContentDirectory';
        this.controlUrl = '/MediaServer/ContentDirectory/Control';
        this.eventSubUrl = '/MediaServer/ContentDirectory/Event';
        this.scpUrl = '/xml/ContentDirectory1.xml';
        /**
         * Default errors and service specific errors
         *
         * @type {SonosUpnpError[]}
         * @remarks See https://svrooij.io/sonos-api-docs/#manual-documentation-file
         */
        this.errors = [
            ...sonos_upnp_errors_1.default.defaultErrors,
            { code: 701, description: 'No such object' },
            { code: 702, description: 'Invalid CurrentTagValue' },
            { code: 703, description: 'Invalid NewTagValue' },
            { code: 704, description: 'Required tag' },
            { code: 705, description: 'Read-only tag' },
            { code: 706, description: 'Parameter mismatch' },
            { code: 708, description: 'Invalid search criteria' },
            { code: 709, description: 'Invalid sort criteria' },
            { code: 710, description: 'No such container' },
            { code: 711, description: 'Restricted object' },
            { code: 712, description: 'Bad metadata' },
            { code: 713, description: 'Restricted parent object' },
            { code: 714, description: 'No such source resource' },
            { code: 715, description: 'Resource access denied' },
            { code: 716, description: 'Transfer busy' },
            { code: 717, description: 'No such file transfer' },
            { code: 718, description: 'No such destination resource' },
            { code: 719, description: 'Destination resource access denied' },
            { code: 720, description: 'Cannot process the request' },
        ];
    }
    // #region actions
    /**
     * Browse for content: Music library (A), share(S:), Sonos playlists(SQ:), Sonos favorites(FV:2), radio stations(R:0/0), radio shows(R:0/1). Recommendation: Send one request, check the &#x60;TotalMatches&#x60; and - if necessary - do additional requests with higher &#x60;StartingIndex&#x60;. In case of duplicates only the first is returned! Example: albums with same title, even if artists are different
     *
     * @param {string} input.ObjectID - The search query, (`A:ARTIST` / `A:ALBUMARTIST` / `A:ALBUM` / `A:GENRE` / `A:COMPOSER` / `A:TRACKS` / `A:PLAYLISTS` / `S:` / `SQ:` / `FV:2` / `R:0/0` / `R:0/1`) with optionally `:search+query` behind it.
     * @param {string} input.BrowseFlag - How to browse [ 'BrowseMetadata' / 'BrowseDirectChildren' ]
     * @param {string} input.Filter - Which fields should be returned `*` for all.
     * @param {number} input.StartingIndex - Paging, where to start, usually 0
     * @param {number} input.RequestedCount - Paging, number of items, maximum is 1,000. This parameter does NOT restrict the number of items being searched (filter) but only the number being returned.
     * @param {string} input.SortCriteria - Sort the results based on metadata fields. `+upnp:artist,+dc:title` for sorting on artist then on title.
     * @remarks (1) If the title contains an apostrophe the returned uri will contain a `&apos;`. (2) Some libraries support a BrowseAndParse, so you don't have to parse the xml.
     */
    async Browse(input) { return await this.SoapRequestWithBody('Browse', input); }
    async CreateObject(input) { return await this.SoapRequestWithBody('CreateObject', input); }
    async DestroyObject(input) { return await this.SoapRequestWithBodyNoResponse('DestroyObject', input); }
    async FindPrefix(input) { return await this.SoapRequestWithBody('FindPrefix', input); }
    async GetAlbumArtistDisplayOption() { return await this.SoapRequest('GetAlbumArtistDisplayOption'); }
    async GetAllPrefixLocations(input) { return await this.SoapRequestWithBody('GetAllPrefixLocations', input); }
    async GetBrowseable() { return await this.SoapRequest('GetBrowseable'); }
    async GetLastIndexChange() { return await this.SoapRequest('GetLastIndexChange'); }
    async GetSearchCapabilities() { return await this.SoapRequest('GetSearchCapabilities'); }
    async GetShareIndexInProgress() { return await this.SoapRequest('GetShareIndexInProgress'); }
    async GetSortCapabilities() { return await this.SoapRequest('GetSortCapabilities'); }
    async GetSystemUpdateID() { return await this.SoapRequest('GetSystemUpdateID'); }
    async RefreshShareIndex(input) { return await this.SoapRequestWithBodyNoResponse('RefreshShareIndex', input); }
    async RequestResort(input) { return await this.SoapRequestWithBodyNoResponse('RequestResort', input); }
    async SetBrowseable(input) { return await this.SoapRequestWithBodyNoResponse('SetBrowseable', input); }
    async UpdateObject(input) { return await this.SoapRequestWithBodyNoResponse('UpdateObject', input); }
    // #endregion
    responseProperties() {
        return {
            Result: 'string',
            NumberReturned: 'number',
            TotalMatches: 'number',
            UpdateID: 'number',
            ObjectID: 'string',
            StartingIndex: 'number',
            AlbumArtistDisplayOption: 'string',
            TotalPrefixes: 'number',
            PrefixAndIndexCSV: 'string',
            IsBrowseable: 'boolean',
            LastIndexChange: 'string',
            SearchCaps: 'string',
            IsIndexing: 'boolean',
            SortCaps: 'string',
            Id: 'number',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            Browseable: 'boolean',
            ContainerUpdateIDs: 'string',
            FavoritePresetsUpdateID: 'string',
            FavoritesUpdateID: 'string',
            RadioFavoritesUpdateID: 'number',
            RadioLocationUpdateID: 'number',
            RecentlyPlayedUpdateID: 'string',
            SavedQueuesUpdateID: 'string',
            SearchCapabilities: 'string',
            ShareIndexInProgress: 'boolean',
            ShareIndexLastError: 'string',
            ShareListUpdateID: 'string',
            SortCapabilities: 'string',
            SystemUpdateID: 'number',
            UserRadioUpdateID: 'string',
        };
    }
}
exports.ContentDirectoryServiceBase = ContentDirectoryServiceBase;
