// @svrooij/sonos/lib/services/device-properties.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevicePropertiesService = void 0;
/**
 * Sonos DeviceProperties service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Modify device properties, like LED status and stereo pairs
 *
 * @export
 * @class DevicePropertiesService
 * @extends {BaseService}
 */
class DevicePropertiesService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'DeviceProperties';
        this.controlUrl = '/DeviceProperties/Control';
        this.eventSubUrl = '/DeviceProperties/Event';
        this.scpUrl = '/xml/DeviceProperties1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async AddBondedZones(input) { return await this.SoapRequestWithBodyNoResponse('AddBondedZones', input); }
    async AddHTSatellite(input) { return await this.SoapRequestWithBodyNoResponse('AddHTSatellite', input); }
    /**
     * Create a stereo pair (left, right speakers), right one becomes hidden
     *
     * @param {string} input.ChannelMapSet - example: `RINCON_B8E9375831C001400:LF,LF;RINCON_000E58FE3AEA01400:RF,RF`
     * @remarks No all speakers support StereoPairs
     */
    async CreateStereoPair(input) { return await this.SoapRequestWithBodyNoResponse('CreateStereoPair', input); }
    async EnterConfigMode(input) { return await this.SoapRequestWithBody('EnterConfigMode', input); }
    async ExitConfigMode(input) { return await this.SoapRequestWithBodyNoResponse('ExitConfigMode', input); }
    async GetAutoplayLinkedZones(input) { return await this.SoapRequestWithBody('GetAutoplayLinkedZones', input); }
    async GetAutoplayRoomUUID(input) { return await this.SoapRequestWithBody('GetAutoplayRoomUUID', input); }
    async GetAutoplayVolume(input) { return await this.SoapRequestWithBody('GetAutoplayVolume', input); }
    /**
     * Get the current button lock state
     */
    async GetButtonLockState() { return await this.SoapRequest('GetButtonLockState'); }
    async GetButtonState() { return await this.SoapRequest('GetButtonState'); }
    async GetHouseholdID() { return await this.SoapRequest('GetHouseholdID'); }
    /**
     * Get the current LED state
     */
    async GetLEDState() { return await this.SoapRequest('GetLEDState'); }
    async GetUseAutoplayVolume(input) { return await this.SoapRequestWithBody('GetUseAutoplayVolume', input); }
    async GetZoneAttributes() { return await this.SoapRequest('GetZoneAttributes'); }
    /**
     * Get information about this specific speaker
     */
    async GetZoneInfo() { return await this.SoapRequest('GetZoneInfo'); }
    async RemoveBondedZones(input) { return await this.SoapRequestWithBodyNoResponse('RemoveBondedZones', input); }
    async RemoveHTSatellite(input) { return await this.SoapRequestWithBodyNoResponse('RemoveHTSatellite', input); }
    async RoomDetectionStartChirping(input) { return await this.SoapRequestWithBody('RoomDetectionStartChirping', input); }
    async RoomDetectionStopChirping(input) { return await this.SoapRequestWithBodyNoResponse('RoomDetectionStopChirping', input); }
    /**
     * Separate a stereo pair
     *
     * @param {string} input.ChannelMapSet - example: `RINCON_B8E9375831C001400:LF,LF;RINCON_000E58FE3AEA01400:RF,RF`
     * @remarks No all speakers support StereoPairs
     */
    async SeparateStereoPair(input) { return await this.SoapRequestWithBodyNoResponse('SeparateStereoPair', input); }
    async SetAutoplayLinkedZones(input) { return await this.SoapRequestWithBodyNoResponse('SetAutoplayLinkedZones', input); }
    async SetAutoplayRoomUUID(input) { return await this.SoapRequestWithBodyNoResponse('SetAutoplayRoomUUID', input); }
    async SetAutoplayVolume(input) { return await this.SoapRequestWithBodyNoResponse('SetAutoplayVolume', input); }
    /**
     * Set the button lock state
     *
     * @param {string} input.DesiredButtonLockState [ 'On' / 'Off' ]
     */
    async SetButtonLockState(input) { return await this.SoapRequestWithBodyNoResponse('SetButtonLockState', input); }
    /**
     * Set the LED state
     *
     * @param {string} input.DesiredLEDState [ 'On' / 'Off' ]
     */
    async SetLEDState(input) { return await this.SoapRequestWithBodyNoResponse('SetLEDState', input); }
    async SetUseAutoplayVolume(input) { return await this.SoapRequestWithBodyNoResponse('SetUseAutoplayVolume', input); }
    async SetZoneAttributes(input) { return await this.SoapRequestWithBodyNoResponse('SetZoneAttributes', input); }
    // #endregion
    responseProperties() {
        return {
            State: 'string',
            IncludeLinkedZones: 'boolean',
            RoomUUID: 'string',
            CurrentVolume: 'number',
            CurrentButtonLockState: 'string',
            CurrentHouseholdID: 'string',
            CurrentLEDState: 'string',
            UseVolume: 'boolean',
            CurrentZoneName: 'string',
            CurrentIcon: 'string',
            CurrentConfiguration: 'string',
            SerialNumber: 'string',
            SoftwareVersion: 'string',
            DisplaySoftwareVersion: 'string',
            HardwareVersion: 'string',
            IPAddress: 'string',
            MACAddress: 'string',
            CopyrightInfo: 'string',
            ExtraInfo: 'string',
            HTAudioIn: 'number',
            Flags: 'number',
            PlayId: 'number',
            ChirpIfPlayingSwappableAudio: 'boolean',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            AirPlayEnabled: 'boolean',
            AutoplayIncludeLinkedZones: 'boolean',
            AutoplayRoomUUID: 'string',
            AutoplaySource: 'string',
            AutoplayUseVolume: 'boolean',
            AutoplayVolume: 'number',
            AvailableRoomCalibration: 'string',
            BehindWifiExtender: 'number',
            ButtonLockState: 'string',
            ChannelFreq: 'number',
            ChannelMapSet: 'string',
            ConfigMode: 'string',
            Configuration: 'string',
            CopyrightInfo: 'string',
            DisplaySoftwareVersion: 'string',
            ExtraInfo: 'string',
            Flags: 'number',
            HardwareVersion: 'string',
            HasConfiguredSSID: 'boolean',
            HdmiCecAvailable: 'boolean',
            HouseholdID: 'string',
            HTAudioIn: 'number',
            HTBondedZoneCommitState: 'number',
            HTFreq: 'number',
            HTSatChanMapSet: 'string',
            Icon: 'string',
            Invisible: 'boolean',
            IPAddress: 'string',
            IsIdle: 'boolean',
            IsZoneBridge: 'boolean',
            KeepGrouped: 'boolean',
            LastChangedPlayState: 'string',
            LEDState: 'string',
            MACAddress: 'string',
            MicEnabled: 'number',
            MoreInfo: 'string',
            Orientation: 'number',
            RoomCalibrationState: 'number',
            SatRoomUUID: 'string',
            SecureRegState: 'number',
            SerialNumber: 'string',
            SettingsReplicationState: 'string',
            SoftwareVersion: 'string',
            SupportsAudioClip: 'boolean',
            SupportsAudioIn: 'boolean',
            TargetRoomName: 'string',
            TVConfigurationError: 'boolean',
            VoiceConfigState: 'number',
            WifiEnabled: 'boolean',
            WirelessLeafOnly: 'boolean',
            WirelessMode: 'number',
            ZoneName: 'string',
        };
    }
}
exports.DevicePropertiesService = DevicePropertiesService;
