// @svrooij/sonos/lib/services/group-rendering-control.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupRenderingControlService = void 0;
/**
 * Sonos GroupRenderingControl service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Volume related controls for groups
 *
 * @export
 * @class GroupRenderingControlService
 * @extends {BaseService}
 */
class GroupRenderingControlService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'GroupRenderingControl';
        this.controlUrl = '/MediaRenderer/GroupRenderingControl/Control';
        this.eventSubUrl = '/MediaRenderer/GroupRenderingControl/Event';
        this.scpUrl = '/xml/GroupRenderingControl1.xml';
        /**
         * Default errors and service specific errors
         *
         * @type {SonosUpnpError[]}
         * @remarks See https://svrooij.io/sonos-api-docs/#manual-documentation-file
         */
        this.errors = [
            ...sonos_upnp_errors_1.default.defaultErrors,
            { code: 701, description: 'Player isn&#x27;t the coordinator' },
        ];
    }
    // #region actions
    /**
     * Get the group mute state.
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @remarks Should be send to coordinator only
     */
    async GetGroupMute(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetGroupMute', input); }
    /**
     * Get the group volume.
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @remarks Should be send to coordinator only
     */
    async GetGroupVolume(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetGroupVolume', input); }
    /**
     * (Un-/)Mute the entire group
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {boolean} input.DesiredMute
     * @remarks Should be send to coordinator only
     */
    async SetGroupMute(input) { return await this.SoapRequestWithBodyNoResponse('SetGroupMute', input); }
    /**
     * Change group volume. Players volume will be changed proportionally based on last snapshot
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {number} input.DesiredVolume - New volume between 0 and 100
     * @remarks Should be send to coordinator only
     */
    async SetGroupVolume(input) { return await this.SoapRequestWithBodyNoResponse('SetGroupVolume', input); }
    /**
     * Relatively change group volume - returns final group volume. Players volume will be changed proportionally based on last snapshot
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {number} input.Adjustment - Number between -100 and +100
     * @remarks Should be send to coordinator only
     */
    async SetRelativeGroupVolume(input) { return await this.SoapRequestWithBody('SetRelativeGroupVolume', input); }
    /**
     * Creates a new group volume snapshot,  the volume ratio between all players. It is used by SetGroupVolume and SetRelativeGroupVolume
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @remarks Should be send to coordinator only
     */
    async SnapshotGroupVolume(input = { InstanceID: 0 }) { return await this.SoapRequestWithBodyNoResponse('SnapshotGroupVolume', input); }
    // #endregion
    responseProperties() {
        return {
            CurrentMute: 'boolean',
            CurrentVolume: 'number',
            NewVolume: 'number',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            GroupMute: 'boolean',
            GroupVolume: 'number',
            GroupVolumeChangeable: 'boolean',
        };
    }
}
exports.GroupRenderingControlService = GroupRenderingControlService;
