// @svrooij/sonos/lib/services/group-management.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupManagementService = void 0;
/**
 * Sonos GroupManagement service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Services related to groups
 *
 * @export
 * @class GroupManagementService
 * @extends {BaseService}
 */
class GroupManagementService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'GroupManagement';
        this.controlUrl = '/GroupManagement/Control';
        this.eventSubUrl = '/GroupManagement/Event';
        this.scpUrl = '/xml/GroupManagement1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async AddMember(input) { return await this.SoapRequestWithBody('AddMember', input); }
    async RemoveMember(input) { return await this.SoapRequestWithBodyNoResponse('RemoveMember', input); }
    async ReportTrackBufferingResult(input) { return await this.SoapRequestWithBodyNoResponse('ReportTrackBufferingResult', input); }
    async SetSourceAreaIds(input) { return await this.SoapRequestWithBodyNoResponse('SetSourceAreaIds', input); }
    // #endregion
    responseProperties() {
        return {
            CurrentTransportSettings: 'string',
            CurrentURI: 'string',
            GroupUUIDJoined: 'string',
            ResetVolumeAfter: 'boolean',
            VolumeAVTransportURI: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            GroupCoordinatorIsLocal: 'boolean',
            LocalGroupUUID: 'string',
            ResetVolumeAfter: 'boolean',
            SourceAreaIds: 'string',
            VirtualLineInGroupID: 'string',
            VolumeAVTransportURI: 'string',
        };
    }
}
exports.GroupManagementService = GroupManagementService;
