// @svrooij/sonos/lib/services/ht-control.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTControlService = void 0;
/**
 * Sonos HTControl service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Service related to the TV remote control
 *
 * @export
 * @class HTControlService
 * @extends {BaseService}
 */
class HTControlService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'HTControl';
        this.controlUrl = '/HTControl/Control';
        this.eventSubUrl = '/HTControl/Event';
        this.scpUrl = '/xml/HTControl1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async CommitLearnedIRCodes(input) { return await this.SoapRequestWithBodyNoResponse('CommitLearnedIRCodes', input); }
    async GetIRRepeaterState() { return await this.SoapRequest('GetIRRepeaterState'); }
    async GetLEDFeedbackState() { return await this.SoapRequest('GetLEDFeedbackState'); }
    async IdentifyIRRemote(input) { return await this.SoapRequestWithBodyNoResponse('IdentifyIRRemote', input); }
    async IsRemoteConfigured() { return await this.SoapRequest('IsRemoteConfigured'); }
    async LearnIRCode(input) { return await this.SoapRequestWithBodyNoResponse('LearnIRCode', input); }
    async SetIRRepeaterState(input) { return await this.SoapRequestWithBodyNoResponse('SetIRRepeaterState', input); }
    async SetLEDFeedbackState(input) { return await this.SoapRequestWithBodyNoResponse('SetLEDFeedbackState', input); }
    // #endregion
    responseProperties() {
        return {
            CurrentIRRepeaterState: 'string',
            LEDFeedbackState: 'string',
            RemoteConfigured: 'boolean',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            IRRepeaterState: 'string',
            LEDFeedbackState: 'string',
            RemoteConfigured: 'boolean',
            TOSLinkConnected: 'boolean',
        };
    }
}
exports.HTControlService = HTControlService;
