// @svrooij/sonos/lib/services/music-services.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MusicServicesServiceBase = void 0;
/**
 * Sonos MusicServices service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Access to external music services, like Spotify or Youtube Music
 *
 * @export
 * @class MusicServicesServiceBase
 * @extends {BaseService}
 */
class MusicServicesServiceBase extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'MusicServices';
        this.controlUrl = '/MusicServices/Control';
        this.eventSubUrl = '/MusicServices/Event';
        this.scpUrl = '/xml/MusicServices1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async GetSessionId(input) { return await this.SoapRequestWithBody('GetSessionId', input); }
    /**
     * Load music service list as xml
     * @remarks Some libraries also support ListAndParseAvailableServices
     */
    async ListAvailableServices() { return await this.SoapRequest('ListAvailableServices'); }
    async UpdateAvailableServices() { return await this.SoapRequestNoResponse('UpdateAvailableServices'); }
    // #endregion
    responseProperties() {
        return {
            SessionId: 'string',
            AvailableServiceDescriptorList: 'string',
            AvailableServiceTypeList: 'string',
            AvailableServiceListVersion: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            ServiceId: 'number',
            ServiceListVersion: 'string',
            SessionId: 'string',
            Username: 'string',
        };
    }
}
exports.MusicServicesServiceBase = MusicServicesServiceBase;
