// @svrooij/sonos/lib/services/music-services.service.extension.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MusicServicesService = void 0;
const array_helper_1 = __importDefault(require("../helpers/array-helper"));
const xml_helper_1 = __importDefault(require("../helpers/xml-helper"));
const music_services_service_1 = require("./music-services.service");
class MusicServicesService extends music_services_service_1.MusicServicesServiceBase {
    /**
     * A parsed version of ListAvailableServices
     *
     * @param {boolean} [cache=false] Should the list be fetched once and then kept in memory?
     * @returns {Promise<Array<MusicService>>} All available music services (not only subscribed ones).
     * @memberof MusicServicesService
     */
    async ListAndParseAvailableServices(cache = false) {
        if (cache === true && this.musicServices !== undefined) {
            return this.musicServices;
        }
        const encodedResponse = await this.ListAvailableServices();
        const raw = xml_helper_1.default.DecodeAndParseXml(encodedResponse.AvailableServiceDescriptorList, '');
        const result = array_helper_1.default.ForceArray(raw.Services.Service)
            .map((service) => MusicServicesService.ParseMusicService(service))
            .sort((a, b) => a.Name.localeCompare(b.Name));
        if (cache === true) {
            this.musicServices = result;
        }
        return result;
    }
    static ParseMusicService(service) {
        var _a, _b;
        const temp = service;
        const result = {
            Id: parseInt(service.Id, 10),
            Policy: {
                Auth: (_a = service.Policy) === null || _a === void 0 ? void 0 : _a.Auth,
                PollInterval: parseInt(((_b = service.Policy) === null || _b === void 0 ? void 0 : _b.PollInterval) || '-1', 10),
            },
        };
        delete temp.Id;
        delete temp.Policy;
        return Object.assign(Object.assign({}, temp), result);
    }
}
exports.MusicServicesService = MusicServicesService;
