// @svrooij/sonos/lib/services/q-play.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QPlayService = void 0;
/**
 * Sonos QPlay service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Services related to Chinese Tencent Qplay service
 *
 * @export
 * @class QPlayService
 * @extends {BaseService}
 */
class QPlayService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'QPlay';
        this.controlUrl = '/QPlay/Control';
        this.eventSubUrl = '/QPlay/Event';
        this.scpUrl = '/xml/QPlay1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async QPlayAuth(input) { return await this.SoapRequestWithBody('QPlayAuth', input); }
    // #endregion
    responseProperties() {
        return {
            Code: 'string',
            MID: 'string',
            DID: 'string',
        };
    }
    // No properties in service description, throw error on retrieval.
    eventProperties() {
        throw new Error('No event properties in service definition');
    }
}
exports.QPlayService = QPlayService;
