// @svrooij/sonos/lib/services/queue.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueService = void 0;
/**
 * Sonos Queue service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Modify and browse queues
 *
 * @export
 * @class QueueService
 * @extends {BaseService}
 */
class QueueService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'Queue';
        this.controlUrl = '/MediaRenderer/Queue/Control';
        this.eventSubUrl = '/MediaRenderer/Queue/Event';
        this.scpUrl = '/xml/Queue1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async AddMultipleURIs(input) { return await this.SoapRequestWithBody('AddMultipleURIs', input); }
    async AddURI(input) { return await this.SoapRequestWithBody('AddURI', input); }
    async AttachQueue(input) { return await this.SoapRequestWithBody('AttachQueue', input); }
    async Backup() { return await this.SoapRequestNoResponse('Backup'); }
    async Browse(input) { return await this.SoapRequestWithBody('Browse', input); }
    async CreateQueue(input) { return await this.SoapRequestWithBody('CreateQueue', input); }
    async RemoveAllTracks(input) { return await this.SoapRequestWithBody('RemoveAllTracks', input); }
    async RemoveTrackRange(input) { return await this.SoapRequestWithBody('RemoveTrackRange', input); }
    async ReorderTracks(input) { return await this.SoapRequestWithBody('ReorderTracks', input); }
    async ReplaceAllTracks(input) { return await this.SoapRequestWithBody('ReplaceAllTracks', input); }
    async SaveAsSonosPlaylist(input) { return await this.SoapRequestWithBody('SaveAsSonosPlaylist', input); }
    // #endregion
    responseProperties() {
        return {
            FirstTrackNumberEnqueued: 'number',
            NumTracksAdded: 'number',
            NewQueueLength: 'number',
            NewUpdateID: 'number',
            QueueID: 'number',
            QueueOwnerContext: 'string',
            Result: 'string',
            NumberReturned: 'number',
            TotalMatches: 'number',
            UpdateID: 'number',
            AssignedObjectID: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            Curated: 'boolean',
            LastChange: 'string',
            UpdateID: 'number',
        };
    }
}
exports.QueueService = QueueService;
