// @svrooij/sonos/lib/services/rendering-control.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderingControlServiceBase = void 0;
/**
 * Sonos RenderingControl service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Volume related controls
 *
 * @export
 * @class RenderingControlServiceBase
 * @extends {BaseService}
 */
class RenderingControlServiceBase extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'RenderingControl';
        this.controlUrl = '/MediaRenderer/RenderingControl/Control';
        this.eventSubUrl = '/MediaRenderer/RenderingControl/Event';
        this.scpUrl = '/xml/RenderingControl1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    /**
     * Get bass level between -10 and 10
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     */
    async GetBass(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetBass', input); }
    /**
     * Get equalizer value
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {string} input.EQType - Allowed values `DialogLevel` (bool) / `MusicSurroundLevel` (-15/+15) /  `NightMode` (bool) / `SubGain` (-10/+10) / `SurroundEnable` (bool) / `SurroundLevel` (-15/+15) / `SurroundMode` (0 = full, 1 = ambient)
     * @remarks Not all EQ types are available on every speaker
     */
    async GetEQ(input) { return await this.SoapRequestWithBody('GetEQ', input); }
    async GetHeadphoneConnected(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetHeadphoneConnected', input); }
    /**
     * Whether or not Loudness is on
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {string} input.Channel [ 'Master' / 'LF' / 'RF' ]
     */
    async GetLoudness(input) { return await this.SoapRequestWithBody('GetLoudness', input); }
    async GetMute(input) { return await this.SoapRequestWithBody('GetMute', input); }
    async GetOutputFixed(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetOutputFixed', input); }
    async GetRoomCalibrationStatus(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetRoomCalibrationStatus', input); }
    async GetSupportsOutputFixed(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetSupportsOutputFixed', input); }
    /**
     * Get treble
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     */
    async GetTreble(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('GetTreble', input); }
    /**
     * Get volume
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {string} input.Channel [ 'Master' / 'LF' / 'RF' ]
     */
    async GetVolume(input) { return await this.SoapRequestWithBody('GetVolume', input); }
    async GetVolumeDB(input) { return await this.SoapRequestWithBody('GetVolumeDB', input); }
    async GetVolumeDBRange(input) { return await this.SoapRequestWithBody('GetVolumeDBRange', input); }
    async RampToVolume(input) { return await this.SoapRequestWithBody('RampToVolume', input); }
    async ResetBasicEQ(input = { InstanceID: 0 }) { return await this.SoapRequestWithBody('ResetBasicEQ', input); }
    async ResetExtEQ(input) { return await this.SoapRequestWithBodyNoResponse('ResetExtEQ', input); }
    async RestoreVolumePriorToRamp(input) { return await this.SoapRequestWithBodyNoResponse('RestoreVolumePriorToRamp', input); }
    /**
     * Set bass level, between -10 and 10
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {number} input.DesiredBass
     */
    async SetBass(input) { return await this.SoapRequestWithBodyNoResponse('SetBass', input); }
    async SetChannelMap(input) { return await this.SoapRequestWithBodyNoResponse('SetChannelMap', input); }
    /**
     * Set equalizer value for different types
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {string} input.EQType - Allowed values `DialogLevel` (bool) / `MusicSurroundLevel` (-15/+15) /  `NightMode` (bool) / `SubGain` (-10/+10) / `SurroundEnable` (bool) / `SurroundLevel` (-15/+15) / `SurroundMode` (0 = full, 1 = ambient)
     * @param {number} input.DesiredValue - Booleans required `1` for true or `0` for false, rest number as specified
     * @remarks Not supported by all speakers, TV related
     */
    async SetEQ(input) { return await this.SoapRequestWithBodyNoResponse('SetEQ', input); }
    /**
     * Set loudness on / off
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {string} input.Channel [ 'Master' / 'LF' / 'RF' ]
     * @param {boolean} input.DesiredLoudness
     */
    async SetLoudness(input) { return await this.SoapRequestWithBodyNoResponse('SetLoudness', input); }
    async SetMute(input) { return await this.SoapRequestWithBodyNoResponse('SetMute', input); }
    async SetOutputFixed(input) { return await this.SoapRequestWithBodyNoResponse('SetOutputFixed', input); }
    async SetRelativeVolume(input) { return await this.SoapRequestWithBody('SetRelativeVolume', input); }
    async SetRoomCalibrationStatus(input) { return await this.SoapRequestWithBodyNoResponse('SetRoomCalibrationStatus', input); }
    async SetRoomCalibrationX(input) { return await this.SoapRequestWithBodyNoResponse('SetRoomCalibrationX', input); }
    /**
     * Set treble level
     *
     * @param {number} input.InstanceID - InstanceID should always be `0`
     * @param {number} input.DesiredTreble - between -10 and 10
     */
    async SetTreble(input) { return await this.SoapRequestWithBodyNoResponse('SetTreble', input); }
    async SetVolume(input) { return await this.SoapRequestWithBodyNoResponse('SetVolume', input); }
    async SetVolumeDB(input) { return await this.SoapRequestWithBodyNoResponse('SetVolumeDB', input); }
    // #endregion
    responseProperties() {
        return {
            CurrentBass: 'number',
            CurrentValue: 'number',
            CurrentHeadphoneConnected: 'boolean',
            CurrentLoudness: 'boolean',
            CurrentMute: 'boolean',
            CurrentFixed: 'boolean',
            RoomCalibrationEnabled: 'boolean',
            RoomCalibrationAvailable: 'boolean',
            CurrentSupportsFixed: 'boolean',
            CurrentTreble: 'number',
            CurrentVolume: 'number',
            MinValue: 'number',
            MaxValue: 'number',
            RampTime: 'number',
            Bass: 'number',
            Treble: 'number',
            Loudness: 'boolean',
            LeftVolume: 'number',
            RightVolume: 'number',
            NewVolume: 'number',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            AudioDelay: 'string',
            AudioDelayLeftRear: 'string',
            AudioDelayRightRear: 'string',
            Bass: 'number',
            DialogLevel: 'string',
            EQValue: 'number',
            HeadphoneConnected: 'boolean',
            LastChange: 'string',
            Loudness: 'boolean',
            MusicSurroundLevel: 'string',
            Mute: 'ChannelValue<boolean>',
            NightMode: 'boolean',
            OutputFixed: 'boolean',
            PresetNameList: 'string',
            RoomCalibrationAvailable: 'boolean',
            RoomCalibrationCalibrationMode: 'string',
            RoomCalibrationCoefficients: 'string',
            RoomCalibrationEnabled: 'boolean',
            RoomCalibrationID: 'string',
            SpeakerSize: 'number',
            SubCrossover: 'string',
            SubEnabled: 'boolean',
            SubGain: 'string',
            SubPolarity: 'string',
            SupportsOutputFixed: 'boolean',
            SurroundEnabled: 'boolean',
            SurroundLevel: 'string',
            SurroundMode: 'string',
            Treble: 'number',
            Volume: 'ChannelValue<number>',
            VolumeDB: 'number',
        };
    }
}
exports.RenderingControlServiceBase = RenderingControlServiceBase;
