// @svrooij/sonos/lib/services/system-properties.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SystemPropertiesServiceBase = void 0;
/**
 * Sonos SystemProperties service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Manage system-wide settings, mainly account stuff
 *
 * @export
 * @class SystemPropertiesServiceBase
 * @extends {BaseService}
 */
class SystemPropertiesServiceBase extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'SystemProperties';
        this.controlUrl = '/SystemProperties/Control';
        this.eventSubUrl = '/SystemProperties/Event';
        this.scpUrl = '/xml/SystemProperties1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async AddAccountX(input) { return await this.SoapRequestWithBody('AddAccountX', input); }
    async AddOAuthAccountX(input) { return await this.SoapRequestWithBody('AddOAuthAccountX', input); }
    async DoPostUpdateTasks() { return await this.SoapRequestNoResponse('DoPostUpdateTasks'); }
    async EditAccountMd(input) { return await this.SoapRequestWithBodyNoResponse('EditAccountMd', input); }
    async EditAccountPasswordX(input) { return await this.SoapRequestWithBodyNoResponse('EditAccountPasswordX', input); }
    async EnableRDM(input) { return await this.SoapRequestWithBodyNoResponse('EnableRDM', input); }
    async GetRDM() { return await this.SoapRequest('GetRDM'); }
    /**
     * Get a saved string.
     *
     * @param {string} input.VariableName - The key for this variable
     * @remarks Strings are saved in the system with SetString, every speaker should return the same data. Will error when not existing
     */
    async GetString(input) { return await this.SoapRequestWithBody('GetString', input); }
    async GetWebCode(input) { return await this.SoapRequestWithBody('GetWebCode', input); }
    async ProvisionCredentialedTrialAccountX(input) { return await this.SoapRequestWithBody('ProvisionCredentialedTrialAccountX', input); }
    async RefreshAccountCredentialsX(input) { return await this.SoapRequestWithBodyNoResponse('RefreshAccountCredentialsX', input); }
    /**
     * Remove a saved string
     *
     * @param {string} input.VariableName - The key for this variable
     * @remarks Not sure what happens if you call this with a VariableName that doesn't exists.
     */
    async Remove(input) { return await this.SoapRequestWithBodyNoResponse('Remove', input); }
    async RemoveAccount(input) { return await this.SoapRequestWithBodyNoResponse('RemoveAccount', input); }
    async ReplaceAccountX(input) { return await this.SoapRequestWithBody('ReplaceAccountX', input); }
    async ResetThirdPartyCredentials() { return await this.SoapRequestNoResponse('ResetThirdPartyCredentials'); }
    async SetAccountNicknameX(input) { return await this.SoapRequestWithBodyNoResponse('SetAccountNicknameX', input); }
    /**
     * Save a string in the system
     *
     * @param {string} input.VariableName - The key for this variable, use something unique
     * @param {string} input.StringValue
     * @remarks Strings are saved in the system, retrieve values with GetString.
     */
    async SetString(input) { return await this.SoapRequestWithBodyNoResponse('SetString', input); }
    // #endregion
    responseProperties() {
        return {
            AccountUDN: 'string',
            AccountNickname: 'string',
            RDMValue: 'boolean',
            StringValue: 'string',
            WebCode: 'string',
            IsExpired: 'boolean',
            NewAccountUDN: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            CustomerID: 'string',
            ThirdPartyHash: 'string',
            UpdateID: 'number',
            UpdateIDX: 'number',
            VoiceUpdateID: 'number',
        };
    }
}
exports.SystemPropertiesServiceBase = SystemPropertiesServiceBase;
