// @svrooij/sonos/lib/services/virtual-line-in.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualLineInService = void 0;
/**
 * Sonos VirtualLineIn service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
class VirtualLineInService extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'VirtualLineIn';
        this.controlUrl = '/MediaRenderer/VirtualLineIn/Control';
        this.eventSubUrl = '/MediaRenderer/VirtualLineIn/Event';
        this.scpUrl = '/xml/VirtualLineIn1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async Next(input = { InstanceID: 0 }) { return await this.SoapRequestWithBodyNoResponse('Next', input); }
    async Pause(input = { InstanceID: 0 }) { return await this.SoapRequestWithBodyNoResponse('Pause', input); }
    async Play(input) { return await this.SoapRequestWithBodyNoResponse('Play', input); }
    async Previous(input = { InstanceID: 0 }) { return await this.SoapRequestWithBodyNoResponse('Previous', input); }
    async SetVolume(input) { return await this.SoapRequestWithBodyNoResponse('SetVolume', input); }
    async StartTransmission(input) { return await this.SoapRequestWithBody('StartTransmission', input); }
    async Stop(input = { InstanceID: 0 }) { return await this.SoapRequestWithBodyNoResponse('Stop', input); }
    async StopTransmission(input) { return await this.SoapRequestWithBodyNoResponse('StopTransmission', input); }
    // #endregion
    responseProperties() {
        return {
            CurrentTransportSettings: 'string',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            AVTransportURIMetaData: 'Track | string',
            CurrentTrackMetaData: 'Track | string',
            CurrentTransportActions: 'string',
            EnqueuedTransportURIMetaData: 'Track | string',
            LastChange: 'string',
        };
    }
}
exports.VirtualLineInService = VirtualLineInService;
