// @svrooij/sonos/lib/services/zone-group-topology.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneGroupTopologyServiceBase = void 0;
/**
 * Sonos ZoneGroupTopology service
 *
 * Stephan van Rooij
 * https://svrooij.io
 *
 * This file is generated, do not edit manually. https://svrooij.io/sonos-api-docs
 */
const base_service_1 = __importDefault(require("./base-service"));
const sonos_upnp_errors_1 = __importDefault(require("./sonos-upnp-errors"));
/**
 * Zone config stuff, eg getting all the configured sonos zones
 *
 * @export
 * @class ZoneGroupTopologyServiceBase
 * @extends {BaseService}
 */
class ZoneGroupTopologyServiceBase extends base_service_1.default {
    constructor() {
        super(...arguments);
        this.serviceNane = 'ZoneGroupTopology';
        this.controlUrl = '/ZoneGroupTopology/Control';
        this.eventSubUrl = '/ZoneGroupTopology/Event';
        this.scpUrl = '/xml/ZoneGroupTopology1.xml';
        this.errors = sonos_upnp_errors_1.default.defaultErrors;
    }
    // #region actions
    async BeginSoftwareUpdate(input) { return await this.SoapRequestWithBodyNoResponse('BeginSoftwareUpdate', input); }
    async CheckForUpdate(input) { return await this.SoapRequestWithBody('CheckForUpdate', input); }
    /**
     * Get information about the current Zone
     */
    async GetZoneGroupAttributes() { return await this.SoapRequest('GetZoneGroupAttributes'); }
    /**
     * Get all the Sonos groups, (as XML)
     * @remarks Some libraries also support GetParsedZoneGroupState that parses the xml for you.
     */
    async GetZoneGroupState() { return await this.SoapRequest('GetZoneGroupState'); }
    async RegisterMobileDevice(input) { return await this.SoapRequestWithBodyNoResponse('RegisterMobileDevice', input); }
    async ReportAlarmStartedRunning() { return await this.SoapRequestNoResponse('ReportAlarmStartedRunning'); }
    async ReportUnresponsiveDevice(input) { return await this.SoapRequestWithBodyNoResponse('ReportUnresponsiveDevice', input); }
    async SubmitDiagnostics(input) { return await this.SoapRequestWithBody('SubmitDiagnostics', input); }
    // #endregion
    responseProperties() {
        return {
            UpdateItem: 'string',
            CurrentZoneGroupName: 'string',
            CurrentZoneGroupID: 'string',
            CurrentZonePlayerUUIDsInGroup: 'string',
            CurrentMuseHouseholdId: 'string',
            ZoneGroupState: 'Array<ZoneGroup> | string',
            DiagnosticID: 'number',
        };
    }
    // Event properties from service description.
    eventProperties() {
        return {
            AlarmRunSequence: 'string',
            AreasUpdateID: 'string',
            AvailableSoftwareUpdate: 'string',
            DiagnosticID: 'number',
            MuseHouseholdId: 'string',
            NetsettingsUpdateID: 'string',
            SourceAreasUpdateID: 'string',
            ThirdPartyMediaServersX: 'string',
            ZoneGroupID: 'string',
            ZoneGroupName: 'string',
            ZoneGroupState: 'Array<ZoneGroup> | string',
            ZonePlayerUUIDsInGroup: 'string',
        };
    }
}
exports.ZoneGroupTopologyServiceBase = ZoneGroupTopologyServiceBase;
