// @svrooij/sonos/lib/sonos-device-base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Sonos base device with all services defined.
 *
 * Stephan van Rooij
 * https://svrooij.io/sonos-api-docs
 *
 * This file is generated, do not edit manually.
 */
const guid_typescript_1 = require("guid-typescript");
const debug_1 = __importDefault(require("debug"));
const services_1 = require("./services");
/**
 * SonosDeviceBase is auto-generated to link to all available services of your Sonos Device
 *
 * @export
 * @class SonosDeviceBase
 */
class SonosDeviceBase {
    /**
     *Creates an instance of SonosDeviceBase.
     * @param {string} host The IP of the speaker
     * @param {number} [port=1400] The port, always 1400
     * @param {string} [uuid=Guid.create().toString()] The UUID of the speaker like RINCON_macaddres01400, used in some commands.
     * @memberof SonosDeviceBase
     */
    constructor(host, port = 1400, uuid = guid_typescript_1.Guid.create().toString()) {
        this.host = host;
        this.port = port;
        this.uuid = uuid;
    }
    get debug() {
        if (this.debugger === undefined) {
            this.debugger = debug_1.default(`sonos:device:${this.host}`);
        }
        return this.debugger;
    }
    /**
      * Control the sonos alarms and times
      * will be initialized on first use.
      *
      * @readonly
      * @type {AlarmClockService}
      * @memberof SonosDeviceBase
      */
    get AlarmClockService() {
        if (this.alarmclockservice === undefined)
            this.alarmclockservice = new services_1.AlarmClockService(this.host, this.port, this.uuid);
        return this.alarmclockservice;
    }
    /**
      * Control line in
      * will be initialized on first use.
      *
      * @readonly
      * @type {AudioInService}
      * @memberof SonosDeviceBase
      */
    get AudioInService() {
        if (this.audioinservice === undefined)
            this.audioinservice = new services_1.AudioInService(this.host, this.port, this.uuid);
        return this.audioinservice;
    }
    /**
      * Service that controls stuff related to transport (play/pause/next/special urls)
      * will be initialized on first use.
      *
      * @readonly
      * @type {AVTransportService}
      * @memberof SonosDeviceBase
      */
    get AVTransportService() {
        if (this.avtransportservice === undefined)
            this.avtransportservice = new services_1.AVTransportService(this.host, this.port, this.uuid);
        return this.avtransportservice;
    }
    /**
      * Services related to connections and protocols
      * will be initialized on first use.
      *
      * @readonly
      * @type {ConnectionManagerService}
      * @memberof SonosDeviceBase
      */
    get ConnectionManagerService() {
        if (this.connectionmanagerservice === undefined)
            this.connectionmanagerservice = new services_1.ConnectionManagerService(this.host, this.port, this.uuid);
        return this.connectionmanagerservice;
    }
    /**
      * Browse for local content
      * will be initialized on first use.
      *
      * @readonly
      * @type {ContentDirectoryService}
      * @memberof SonosDeviceBase
      */
    get ContentDirectoryService() {
        if (this.contentdirectoryservice === undefined)
            this.contentdirectoryservice = new services_1.ContentDirectoryService(this.host, this.port, this.uuid);
        return this.contentdirectoryservice;
    }
    /**
      * Modify device properties, like LED status and stereo pairs
      * will be initialized on first use.
      *
      * @readonly
      * @type {DevicePropertiesService}
      * @memberof SonosDeviceBase
      */
    get DevicePropertiesService() {
        if (this.devicepropertiesservice === undefined)
            this.devicepropertiesservice = new services_1.DevicePropertiesService(this.host, this.port, this.uuid);
        return this.devicepropertiesservice;
    }
    /**
      * Services related to groups
      * will be initialized on first use.
      *
      * @readonly
      * @type {GroupManagementService}
      * @memberof SonosDeviceBase
      */
    get GroupManagementService() {
        if (this.groupmanagementservice === undefined)
            this.groupmanagementservice = new services_1.GroupManagementService(this.host, this.port, this.uuid);
        return this.groupmanagementservice;
    }
    /**
      * Volume related controls for groups
      * will be initialized on first use.
      *
      * @readonly
      * @type {GroupRenderingControlService}
      * @memberof SonosDeviceBase
      */
    get GroupRenderingControlService() {
        if (this.grouprenderingcontrolservice === undefined)
            this.grouprenderingcontrolservice = new services_1.GroupRenderingControlService(this.host, this.port, this.uuid);
        return this.grouprenderingcontrolservice;
    }
    /**
      * Service related to the TV remote control
      * will be initialized on first use.
      *
      * @readonly
      * @type {HTControlService}
      * @memberof SonosDeviceBase
      */
    get HTControlService() {
        if (this.htcontrolservice === undefined)
            this.htcontrolservice = new services_1.HTControlService(this.host, this.port, this.uuid);
        return this.htcontrolservice;
    }
    /**
      * Access to external music services, like Spotify or Youtube Music
      * will be initialized on first use.
      *
      * @readonly
      * @type {MusicServicesService}
      * @memberof SonosDeviceBase
      */
    get MusicServicesService() {
        if (this.musicservicesservice === undefined)
            this.musicservicesservice = new services_1.MusicServicesService(this.host, this.port, this.uuid);
        return this.musicservicesservice;
    }
    /**
      * Services related to Chinese Tencent Qplay service
      * will be initialized on first use.
      *
      * @readonly
      * @type {QPlayService}
      * @memberof SonosDeviceBase
      */
    get QPlayService() {
        if (this.qplayservice === undefined)
            this.qplayservice = new services_1.QPlayService(this.host, this.port, this.uuid);
        return this.qplayservice;
    }
    /**
      * Modify and browse queues
      * will be initialized on first use.
      *
      * @readonly
      * @type {QueueService}
      * @memberof SonosDeviceBase
      */
    get QueueService() {
        if (this.queueservice === undefined)
            this.queueservice = new services_1.QueueService(this.host, this.port, this.uuid);
        return this.queueservice;
    }
    /**
      * Volume related controls
      * will be initialized on first use.
      *
      * @readonly
      * @type {RenderingControlService}
      * @memberof SonosDeviceBase
      */
    get RenderingControlService() {
        if (this.renderingcontrolservice === undefined)
            this.renderingcontrolservice = new services_1.RenderingControlService(this.host, this.port, this.uuid);
        return this.renderingcontrolservice;
    }
    /**
      * Manage system-wide settings, mainly account stuff
      * will be initialized on first use.
      *
      * @readonly
      * @type {SystemPropertiesService}
      * @memberof SonosDeviceBase
      */
    get SystemPropertiesService() {
        if (this.systempropertiesservice === undefined)
            this.systempropertiesservice = new services_1.SystemPropertiesService(this.host, this.port, this.uuid);
        return this.systempropertiesservice;
    }
    get VirtualLineInService() {
        if (this.virtuallineinservice === undefined)
            this.virtuallineinservice = new services_1.VirtualLineInService(this.host, this.port, this.uuid);
        return this.virtuallineinservice;
    }
    /**
      * Zone config stuff, eg getting all the configured sonos zones
      * will be initialized on first use.
      *
      * @readonly
      * @type {ZoneGroupTopologyService}
      * @memberof SonosDeviceBase
      */
    get ZoneGroupTopologyService() {
        if (this.zonegrouptopologyservice === undefined)
            this.zonegrouptopologyservice = new services_1.ZoneGroupTopologyService(this.host, this.port, this.uuid);
        return this.zonegrouptopologyservice;
    }
    /**
     * Refresh all service event subscriptions, if you were already subscribed.
     *
     * @returns {Promise<boolean>} returns true if at least one service subscription is refreshed.
     * @memberof SonosDeviceBase
     */
    async RefreshEventSubscriptions() {
        let result = false;
        if (this.alarmclockservice !== undefined)
            result = await this.alarmclockservice.CheckEventListener() || result;
        if (this.audioinservice !== undefined)
            result = await this.audioinservice.CheckEventListener() || result;
        if (this.avtransportservice !== undefined)
            result = await this.avtransportservice.CheckEventListener() || result;
        if (this.connectionmanagerservice !== undefined)
            result = await this.connectionmanagerservice.CheckEventListener() || result;
        if (this.contentdirectoryservice !== undefined)
            result = await this.contentdirectoryservice.CheckEventListener() || result;
        if (this.devicepropertiesservice !== undefined)
            result = await this.devicepropertiesservice.CheckEventListener() || result;
        if (this.groupmanagementservice !== undefined)
            result = await this.groupmanagementservice.CheckEventListener() || result;
        if (this.grouprenderingcontrolservice !== undefined)
            result = await this.grouprenderingcontrolservice.CheckEventListener() || result;
        if (this.htcontrolservice !== undefined)
            result = await this.htcontrolservice.CheckEventListener() || result;
        if (this.musicservicesservice !== undefined)
            result = await this.musicservicesservice.CheckEventListener() || result;
        if (this.qplayservice !== undefined)
            result = await this.qplayservice.CheckEventListener() || result;
        if (this.queueservice !== undefined)
            result = await this.queueservice.CheckEventListener() || result;
        if (this.renderingcontrolservice !== undefined)
            result = await this.renderingcontrolservice.CheckEventListener() || result;
        if (this.systempropertiesservice !== undefined)
            result = await this.systempropertiesservice.CheckEventListener() || result;
        if (this.virtuallineinservice !== undefined)
            result = await this.virtuallineinservice.CheckEventListener() || result;
        if (this.zonegrouptopologyservice !== undefined)
            result = await this.zonegrouptopologyservice.CheckEventListener() || result;
        return result;
    }
    GetServiceByName(name) {
        if (typeof name !== 'string' || name === '')
            return undefined;
        const lowerName = name.toLowerCase();
        if (lowerName === 'alarmclockservice') {
            return this.AlarmClockService;
        }
        if (lowerName === 'audioinservice') {
            return this.AudioInService;
        }
        if (lowerName === 'avtransportservice') {
            return this.AVTransportService;
        }
        if (lowerName === 'connectionmanagerservice') {
            return this.ConnectionManagerService;
        }
        if (lowerName === 'contentdirectoryservice') {
            return this.ContentDirectoryService;
        }
        if (lowerName === 'devicepropertiesservice') {
            return this.DevicePropertiesService;
        }
        if (lowerName === 'groupmanagementservice') {
            return this.GroupManagementService;
        }
        if (lowerName === 'grouprenderingcontrolservice') {
            return this.GroupRenderingControlService;
        }
        if (lowerName === 'htcontrolservice') {
            return this.HTControlService;
        }
        if (lowerName === 'musicservicesservice') {
            return this.MusicServicesService;
        }
        if (lowerName === 'qplayservice') {
            return this.QPlayService;
        }
        if (lowerName === 'queueservice') {
            return this.QueueService;
        }
        if (lowerName === 'renderingcontrolservice') {
            return this.RenderingControlService;
        }
        if (lowerName === 'systempropertiesservice') {
            return this.SystemPropertiesService;
        }
        if (lowerName === 'virtuallineinservice') {
            return this.VirtualLineInService;
        }
        if (lowerName === 'zonegrouptopologyservice') {
            return this.ZoneGroupTopologyService;
        }
        return undefined;
    }
}
exports.default = SonosDeviceBase;
