// @svrooij/sonos/lib/sonos-device-discovery.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const dgram_1 = require("dgram");
const debug_1 = __importDefault(require("debug"));
class SonosDiscoveryError extends Error {
    constructor() {
        super('No players found');
    }
}
class SonosDeviceDiscovery {
    constructor() {
        this.devices = [];
        this.events = new events_1.EventEmitter();
        this.isBound = false;
        this.DeviceAvailable = 'DeviceAvailable';
        this.SEARCH_STRING = Buffer.from(['M-SEARCH * HTTP/1.1',
            'HOST: 239.255.255.250:1900',
            'MAN: "ssdp:discover"',
            'MX: 1',
            'ST: urn:schemas-upnp-org:device:ZonePlayer:1',
        ].join('\r\n'));
        this.debug = debug_1.default('sonos:device-discovery');
        this.socket = dgram_1.createSocket({ type: 'udp4', reuseAddr: true });
        this.socket.on('message', (buffer, rinfo) => {
            const msg = buffer.toString();
            this.debug('SSDP message\r\n%s', msg);
            if (/.Sonos.+/.test(msg)) {
                const modelCheck = /SERVER.{0,200}\((.{2,50})\)/.exec(msg);
                const model = (modelCheck && modelCheck.length > 1 ? modelCheck[1] : undefined);
                const addr = rinfo.address;
                if (model && !this.devices.some((d) => d.host === addr)) {
                    this.debug('Found %s on %s', model, addr);
                    const player = { host: addr, port: 1400, model };
                    this.devices.push(player);
                    this.events.emit(this.DeviceAvailable, player);
                }
            }
        });
        this.socket.on('listening', () => {
            this.socket.addMembership('239.255.255.250');
            this.socket.setMulticastTTL(128);
            this.debug('UDP socket started listening');
        });
        // this.socket.on('error', (err) => {
        //   this.debug('Error with socket', err);
        // });
        this.socket.bind(() => {
            this.debug('Bound to port %d', this.socket.address().port);
            this.isBound = true;
        });
        this.debug('Device discovery created');
    }
    get port() {
        var _a;
        return (_a = this.socket) === null || _a === void 0 ? void 0 : _a.address().port;
    }
    sendBroadcast() {
        this.debug('sendBroadcast()');
        if (this.isBound !== true) {
            this.debug('UDP port not bound, waiting 100ms');
            this.pollTimeout = setTimeout(() => {
                this.sendBroadcast();
            }, 100);
            return;
        }
        this.sendBroadcastToAddress('239.255.255.250');
        this.sendBroadcastToAddress('255.255.255.255');
        this.pollTimeout = setTimeout(() => {
            this.sendBroadcast();
        }, 3000);
    }
    sendBroadcastToAddress(addr) {
        // this.debug('sendBroadcastToAddress(\'%s\')', addr);
        const buff = this.SEARCH_STRING;
        this.socket.send(buff, 0, buff.length, 1900, addr);
    }
    setCancelTimeout(timeoutSeconds) {
        this.searchTimeout = setTimeout(() => {
            this.debug('Timeout occured');
            this.events.emit('timeout');
            this.Cancel();
        }, timeoutSeconds * 1000);
    }
    SearchOne(timeoutSeconds = 10) {
        this.debug('Search one device in %d seconds', timeoutSeconds);
        return new Promise((resolve, reject) => {
            this.events.once(this.DeviceAvailable, (player) => {
                this.Cancel();
                resolve(player);
            });
            this.events.once('timeout', () => {
                reject(new SonosDiscoveryError());
            });
            this.setCancelTimeout(timeoutSeconds);
            this.sendBroadcast();
        });
    }
    Search(timeoutSeconds = 30) {
        this.debug('Search all devices in %d seconds', timeoutSeconds);
        return new Promise((resolve, reject) => {
            this.events.once('timeout', () => {
                if (this.devices.length > 0) {
                    resolve(this.devices);
                }
                else {
                    reject(new SonosDiscoveryError());
                }
            });
            this.setCancelTimeout(timeoutSeconds);
            this.sendBroadcast();
        });
    }
    Cancel() {
        if (this.pollTimeout !== undefined) {
            clearTimeout(this.pollTimeout);
        }
        if (this.searchTimeout !== undefined) {
            clearTimeout(this.searchTimeout);
        }
        if (this.socket) {
            this.socket.close();
        }
    }
}
exports.default = SonosDeviceDiscovery;
