// @svrooij/sonos/lib/helpers/tts-helper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_fetch_1 = __importStar(require("node-fetch"));
const debug_1 = __importDefault(require("debug"));
const http_error_1 = __importDefault(require("../models/http-error"));
class TtsHelper {
    static async GetTtsUriFromEndpoint(endpoint, text, language, gender, name, engine) {
        TtsHelper.debug('Getting tts uri from server %s', endpoint);
        const request = new node_fetch_1.Request(endpoint, {
            method: 'POST',
            headers: {
                'Content-type': 'application/json',
                'User-Agent': `node-sonos-ts/${TtsHelper.pack.version} (${TtsHelper.pack.homepage})`,
            },
            body: JSON.stringify({
                text, lang: language, gender, name, engine,
            }),
        });
        const response = await node_fetch_1.default(request);
        if (!response.ok) {
            this.debug('handleRequest error %d %s', response.status, response.statusText);
            throw new http_error_1.default('GetTtsUriFromEndpoint', response.status, response.statusText);
        }
        const data = JSON.parse(await response.text());
        return data.cdnUri || data.uri;
    }
    static async TtsOptionsToNotification(options) {
        var _a, _b;
        const endpoint = (_a = options.endpoint) !== null && _a !== void 0 ? _a : process.env.SONOS_TTS_ENDPOINT;
        if (endpoint === undefined) {
            throw new Error('TTS endpoint is required, check the documentation.');
        }
        const lang = options.lang || process.env.SONOS_TTS_LANG;
        if (lang === undefined || lang === '') {
            throw new Error('TTS lang is required.');
        }
        if (options.text === '' || options.text === undefined || options.text.length < 2) {
            throw new Error('TTS text is required, duh!');
        }
        const engine = options.engine || process.env.SONOS_TTS_ENGINE;
        const uri = await TtsHelper.GetTtsUriFromEndpoint(endpoint, options.text, lang, options.gender, options.name, engine);
        // Typescript way to convert objects, someone got a better way?
        const notificationOptions = options;
        notificationOptions.trackUri = uri;
        notificationOptions.timeout = (_b = notificationOptions.timeout) !== null && _b !== void 0 ? _b : 60;
        return notificationOptions;
    }
}
exports.default = TtsHelper;
TtsHelper.debug = debug_1.default('sonos:tts');
// TODO Automatically get from package on build
TtsHelper.pack = { version: '1.1.5', homepage: 'https://github.com/svrooij/node-sonos-ts' };
