// @svrooij/sonos/lib/helpers/xml-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fast_xml_parser_1 = require("fast-xml-parser");
const html_entities_1 = require("html-entities");
class XmlHelper {
    /**
     * Decode an encoded xml string
     *
     * @static
     * @param {string} text Encoded XML
     * @returns {string} Decoded XML
     * @memberof XmlHelper
     */
    static DecodeXml(text) {
        if (typeof text !== 'string' || text === '') {
            return undefined;
        }
        return html_entities_1.decode(text, { level: 'xml' });
    }
    /**
     * Decode an encoded xml string
     *
     * @static
     * @param {string} text Encoded XML
     * @returns {string} Decoded XML
     * @memberof XmlHelper
     */
    static DecodeHtml(text) {
        if (typeof text === 'undefined' || (typeof text === 'string' && text === '')) {
            return undefined;
        }
        if (typeof text !== 'string') {
            return XmlHelper.DecodeHtml(`${text}`);
        }
        return html_entities_1.decode(text);
    }
    /**
     * DecodeAndParseXml will decode the encoded xml string and then try to parse it
     *
     * @static
     * @param {string} encodedXml Encoded Xml string
     * @returns {*} a parsed Object of the XML string
     * @memberof XmlHelper
     */
    static DecodeAndParseXml(encodedXml, attributeNamePrefix = '_') {
        const decoded = XmlHelper.DecodeXml(encodedXml);
        if (typeof decoded === 'undefined')
            return undefined;
        return fast_xml_parser_1.parse(decoded, { ignoreAttributes: false, attributeNamePrefix });
    }
    /**
     * DecodeAndParseXml will decode the encoded xml string and then try to parse it
     *
     * @static
     * @param {string} encodedXml Encoded Xml string
     * @returns {*} a parsed Object of the XML string
     * @memberof XmlHelper
     */
    static DecodeAndParseXmlNoNS(encodedXml, attributeNamePrefix = '_') {
        const decoded = XmlHelper.DecodeXml(encodedXml);
        return decoded ? fast_xml_parser_1.parse(decoded, { ignoreAttributes: false, ignoreNameSpace: true, attributeNamePrefix }) : undefined;
    }
    /**
     * EncodeXml will encode a xml string so it is safe to send to sonos.
     *
     * @static
     * @param {string} xml
     * @returns {string}
     * @memberof XmlHelper
     */
    static EncodeXml(xml) {
        if (typeof xml !== 'string' || xml === '')
            return '';
        return html_entities_1.encode(xml, { level: 'xml' });
    }
    static EncodeTrackUri(trackUri) {
        if (trackUri.startsWith('http'))
            return encodeURI(trackUri);
        if (trackUri.startsWith('x-sonos-hta')
            || trackUri.startsWith('x-rincon-mp3radio'))
            return trackUri;
        if (trackUri.startsWith('x-rincon-playlist:')) {
            const index = trackUri.indexOf('/');
            return trackUri.substr(0, index) + this.EncodeXml(trackUri.substr(index)).replace(/:/g, '%3a').replace(/ /g, '%20');
        }
        // Part below needs some work.
        const index = trackUri.indexOf(':') + 1;
        return trackUri.substr(0, index) + this.EncodeXml(trackUri.substr(index)).replace(/:/g, '%3a');
    }
    static DecodeTrackUri(input) {
        if (typeof input !== 'string' || input === '') {
            return undefined;
        }
        return XmlHelper.DecodeXml(decodeURIComponent(input));
    }
}
exports.default = XmlHelper;
