// @svrooij/sonos/lib/models/notificationQueue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationQueue = exports.NotificationQueueTimeoutItem = void 0;
class NotificationQueueTimeoutItem {
    constructor(
    /**
    * The timeout reference to clear if anything is okay
    *
    * @type {NotificationQueueTimeoutItem}
    */
    timeout, 
    /**
    * The timestamp when the timeout will fire
    *
    * @type {NotificationQueueTimeoutItem}
    */
    fireTime) {
        this.timeout = timeout;
        this.fireTime = fireTime;
    }
    timeLeft() {
        return this.fireTime - (new Date()).getTime();
    }
}
exports.NotificationQueueTimeoutItem = NotificationQueueTimeoutItem;
class NotificationQueue {
    constructor() {
        this.queue = [];
        this.promisesToResolve = [];
        this.playing = false;
        /**
        * Whether any item in the Queue changed the volume
        *
        * @type {boolean}
        */
        this.volumeChanged = false;
        /**
        * Whether the Queue played any item at all
        *
        * @type {boolean}
        */
        this.anythingPlayed = false;
    }
}
exports.NotificationQueue = NotificationQueue;
