// @svrooij/sonos/lib/models/sonos-events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SonosEvents = void 0;
var SonosEvents;
(function (SonosEvents) {
    /**
     * Unparsed events from the AVTransport service
     */
    SonosEvents["AVTransport"] = "avtransport";
    SonosEvents["CurrentTrackUri"] = "currentTrackUri";
    SonosEvents["CurrentTrackMetadata"] = "currentTrack";
    SonosEvents["EnqueuedTransportUri"] = "enqueuedTransportUri";
    SonosEvents["EnqueuedTransportMetadata"] = "enqueuedTransport";
    SonosEvents["NextTrackUri"] = "nextTrackUri";
    SonosEvents["NextTrackMetadata"] = "nextTrack";
    /**
     * Changes in TransportState
     */
    SonosEvents["CurrentTransportState"] = "transportState";
    SonosEvents["CurrentTransportStateSimple"] = "simpleTransportState";
    /**
     * If the CurrentTransportState changed to STOPPED
     */
    SonosEvents["PlaybackStopped"] = "playbackStopped";
    /**
     * Unparsed events from the RenderingControl service
     */
    SonosEvents["RenderingControl"] = "renderingcontrol";
    SonosEvents["Mute"] = "muted";
    SonosEvents["Volume"] = "volume";
    /**
     * This event is emitted if the coordinator of this device changed
     */
    SonosEvents["Coordinator"] = "coordinator";
    /**
     * This event is emitted if the groupname changes.
     */
    SonosEvents["GroupName"] = "groupname";
    /**
     * This event is emitted if something went wrong while subscribing for events
     */
    SonosEvents["SubscriptionError"] = "subscriptionError";
})(SonosEvents = exports.SonosEvents || (exports.SonosEvents = {}));
